modules/mono/csharp_script.cpp

```diff
@@ -2662,13 +2666,15 @@ bool CSharpScript::has_script_signal(const StringName &p_signal) const {
 	return false;
 }
 
-void CSharpScript::get_script_signal_list(List<MethodInfo> *r_signals) const {
+void CSharpScript::get_script_signal_list(List<MethodInfo> *r_signals, bool p_no_inheritance) const {
 	if (!valid) {
 		return;
 	}
 
 	for (const EventSignalInfo &signal : get_script_event_signals()) {
-		r_signals->push_back(signal.method_info);
+		if (!p_no_inheritance || signal.method_info.class_name == class_name.operator StringName()) {
+			r_signals->push_back(signal.method_info);
+		}
```

Comment:
Note: I haven't tested this.

Reply:
Unlike `PropertyInfo` `MethodInfo` does not have a `class_name` member

Reply:
I think the info about the signal-class relationship should be added here. But I hope that C# colleagues will help me with this, since I am not familiar with the code.

https://github.com/godotengine/godot/blob/f7bc653cbe81018fe362472a0143b7153a52f929/modules/mono/csharp_script.cpp#L2280

https://github.com/godotengine/godot/blob/f7bc653cbe81018fe362472a0143b7153a52f929/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs#L598-L602

https://github.com/godotengine/godot/blob/f7bc653cbe81018fe362472a0143b7153a52f929/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs#L645-L685

Reply:
**TL;DR:** Sorry, I'm not sure this can be made to work in C# currently. Making it work would probably require making a lot of changes that would grow the scope of the PR.

---

IIRC the members (properties, methods and signals) in `CSharpScript` include inherited members, so even if you don't iterate the script inheritance you'll still get the inherited members.

What I mean is that `script->methods` includes the methods from the current script and its base script, so the loop in `get_script_method_list` that iterates the inheritance chain actually returns duplicates (see https://github.com/godotengine/godot/pull/77273).

The linked PR is trying to fix this by only retrieving the members from the current script and not inherited ones, then the loop would make sense. The problem with this approach is that some of the hierarchy chain may not be represented as a `CSharpScript` (e.g.: generic types) and so we would be unable to retrieve members from those types.

The alternative approach would be to remove the loop and keep storing all the members (including inherited ones) in the script. This is what I suggested in the linked PR. However, I don't think it'd be possible to differentiate between members that belong to the current script and inherited members, so it would be impossible to implement `p_no_inherited`.

I think the better solution would be the first approach (each script only contains its own members and not inherited ones) but then we'd have to fix the script hierarchy chain and start creating `CSharpScript` for currently unrepresentable types. I don't think it's possible to create a `CSharpScript` for a generic type because we don't know what the generic type arguments are when the generic type is unbound (what type is `T`?) so we can't retrieve the members from the type. We could create `CSharpScript` only for bound generic types (where we know what `T` is), but the problem is that then we'd need to create multiple `CSharpScript` with the same resource path and Godot doesn't like that.

So, as things currently are, I think it may not be possible to implement this without fixing the current issues with `CSharpScript` and that would be out of scope for this PR. Would it be OK to ignore `p_no_inheritance` for now? Or would that create issues?

Reply:
> Would it be OK to ignore `p_no_inheritance` for now? Or would that create issues?

This is required by #80411 so that the script hierarchy is displayed correctly in the Node > Signals tab. Now, to do this, I get a list of all signals of the current class and exclude from it all signals of the parent class. If C# will ignore `p_no_inheritance`, then the signal will be duplicated in the Signals tab, which is not good.

As far as I understand from your explanation, C# reflection does not allow you to get a list of members defined only in the current class, without base classes (plus C# has interfaces and maybe some other forms of multiple inheritance). But could we use the approach I described above (exclude members of parent script)? Yes, the first script can still include members that don't belong to it, but at least inherited scripts won't have duplicates.

Reply:
> This is required by https://github.com/godotengine/godot/pull/80411 so that the script hierarchy is displayed correctly in the Node > Signals tab.

Well, there are already other similar problems. As you can see reported in https://github.com/godotengine/godot/pull/77273, the connect dialog shows duplicate methods. So if that's the only problem, maybe we can just fix it in a follow-up?

>  C# reflection does not allow you to get a list of members defined only in the current class, without base classes

Just to clarify, C# reflection allows it, we're just not currently doing that in `UpdateScriptClassInfo`. https://github.com/godotengine/godot/pull/77273 intends to fix this.

> But could we use the approach I described above (exclude members of parent script)?

Sorry I think I've missed this, where is this approach described?

Do you mean including the name of the class that declares the member in `MethodInfo` and using that to check? If so, this would require making changes to Core which greatly increases the scope of your PR. I think I'd rather wait for https://github.com/godotengine/godot/pull/77273 to fix the issue properly.

By the way, `PropertyInfo::class_name` has been mentioned before, this is not the name of the class that declares the property, it's the name of the class that is the property's type (if that type is a class).

Reply:
> Sorry I think I've missed this, where is this approach described?

I mean something like [this](https://github.com/godotengine/godot/blob/206be5dec72a3e67b8bcd758eb86d60b67d46fa2/editor/connections_dialog.cpp#L1286-L1300).

> Well, there are already other similar problems. As you can see reported in #77273, the connect dialog shows duplicate methods. So if that's the only problem, maybe we can just fix it in a follow-up?

OK, I added TODOs.

Reply:
> I mean something like [this](https://github.com/godotengine/godot/blob/206be5dec72a3e67b8bcd758eb86d60b67d46fa2/editor/connections_dialog.cpp#L1286-L1300).

Oh yeah, that's smart.